### PR TITLE
Avoid using recursion when building tracebacks

### PR DESCRIPTION
**Pull request by richvdh**
The list of frames in a Failure may be very long - particularly if the failure
was due to a stack overflow. It is unhelpful if our error-handling functions
themselves fail with stack overflows when handling a stack overflow situation.

To fix one particular failure mode, avoid using recursion to handle the frames
list.

Fixes https://twistedmatrix.com/trac/ticket/9301.

### CODE REVIEW

**Comment by richvdh** [comment]
I don't *think* the CI failures on python 3.6 are my doing, but feel free to tell me how I'm wrong...

**Comment by markrwilliams** [comment]
@richvdh Definitely not your fault - @rodrigc [recently fixed this](https://github.com/twisted/twisted/commit/e58956fe12af7b41f6c1735ba7b14cba69e82a5e).

File: src/twisted/python/failure.py [hunk 1]
```diff
@@ -90,30 +90,41 @@ class NoCurrentExceptionError(Exception):
 
 
 
-class _Traceback(object):
+def _Traceback(frames):
```
Comment (markrwilliams) [inline src/twisted/python/failure.py:93]: I'm a little worried about switching this to a function, but it's a private API and it doesn't seem to be called anywhere outside of `twisted.python.failure`.

**Review by markrwilliams — APPROVED** [review]
This looks reasonable - thanks for the contribution!

This is ready to merge when the builds succeed.

**Pull request merged**